### PR TITLE
Polio 1291:  Vaccine Supply chain, details view front-end

### PR DESCRIPTION
**Pull request by quang-le**
Details view of polio vaccine supply chain

Related JIRA tickets : POLIO-1291

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- Small improvement to rerouting to user/plugin homepage
- Add details page
    - each tab can be saved separately
    - option to save all
    - updates are managed through multiple PATCH, POST, DELETE requests, i.o 1 big PUT request like polio campaigns page 
- Add `MultiDateCell` and `MultiDateTimeCellRfc` for cleaner display of multiple dates in table cells
- Add `useSkipEffectUntilValue` hook, to block `useEffect` trigger before data is fetched
-  Fix some bugs in `TextArea` component (related to debounce feature)
- Add more generic `TabWithInfoIcon` component, derived from `PolioTab`
- Improve `useObjectState` hook (handle some edge cases)
- Add `useTab` hook, to handle `Tab` state and change handler
- removed `isTouched` function from utils. Replaced with more reliable `get` from lodash
- Add `shrinkLabel` prop to polio's `TextInput` to make UI look more consistent
- Remove vaccine tab related code in translations and validation
- update `useGetCampaigns` to enable passing options to the underlying `useQuery`hook

## How to test

*Setup*:
Enable `SHOW_DEV_FEATURES` feature flag on your local DB polio account

*Tests*:
- Go to VRF list. 
- Click create:
    - Pre-alerts and VAR tabs should be disabled
    - Fill in fields
    - click cancel: fields should be cleared
    - Fill in fields again and save:
        - Pre-alerts and VAR tabs should be active
        - The top bar title should be updated
        - There should be a "Save All" button
   - go to either PreAlerts or VAR:
       - click add button
       - clicking the delete before saving should remove the pre-alert/var from the UI
       - Go to VRF tab and change a field without saving
       - Go to prealert and save the prealert(s)
       - Go back to VRf tab: the change made before saving the prealert should not have overwritten
       - Click cancel: the change in the vrf should be reverted

Try variations of this flow with:
- Using save all when updating 1,2 or 3 tabs
- Saving one tab, then saving another, saving all, or canceling
- In pre-alerts and VAR: try adding, updating and deleting entries with a single save 

Note: 
- When saving multiple tabs, each tab should have its own snackbar if update is successful
- When saving, if there is an error, each call will have it's own error snackbar (create, update, delete)

## Print screen / video

https://github.com/BLSQ/iaso/assets/38907762/e55187bf-5611-4a46-8688-4bab244c63c9

## Notes

Merges into the backend branch https://github.com/BLSQ/iaso/tree/POLIO-1301-vaccine-supply-chain-backend-part-2

failing python tests will be addressed there

### CODE REVIEW

File: hat/assets/js/apps/Iaso/domains/forms/requests.js [hunk 1]
```diff
@@ -15,8 +15,3 @@ export const useGetForm = formId =>
             enabled: formId && formId !== '0',
         },
     );
-
-export const useRefreshForm = formId => {
-    const queryClient = useQueryClient();
-    return data => queryClient.setQueryData(['forms', formId], data);
-};
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/forms/requests.js:23]: This wasn't used anywhere

File: hat/assets/js/apps/Iaso/hooks/useTabs.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            };
+            dispatch(redirectToReplace(baseUrl, newParams));
+            setTab(newTab);
+        },
+        [params, dispatch, baseUrl],
+    );
+    return useMemo(
+        () => ({ tab, setTab, handleChangeTab }),
+        [handleChangeTab, tab],
+    );
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/hooks/useTabs.tsx:48]: ❤️

File: plugins/polio/js/src/components/Inputs/MultiSelect.tsx [hunk 3]
```diff
@@ -34,12 +39,17 @@ export const MultiSelect: FunctionComponent<Props> = ({
...
-                form.setFieldTouched(field.name, true);
-                form.setFieldValue(field.name, value);
+            onChange={(keyValue, value) => {
+                if (onChange) {
+                    onChange(keyValue, value);
+                } else {
+                    form.setFieldTouched(field.name, true);
+                    form.setFieldValue(field.name, value);
+                }
             }}
```
Comment (beygorghor) [inline plugins/polio/js/src/components/Inputs/MultiSelect.tsx:53]:
```
onChange?.(keyValue, value) || (() => {
    form.setFieldTouched(field.name, true);
    form.setFieldValue(field.name, value);
})();
```

Comment (beygorghor) [reply plugins/polio/js/src/components/Inputs/MultiSelect.tsx:53]: ❓

File: plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VaccineRequestForm/VaccineRequestForm.tsx [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    // }, [setFieldValue, values?.vrf?.country]);
+
+    // useSkipEffectOnMount(() => {
+    //     setFieldValue('vrf.vaccine_type', undefined);
+    //     setFieldValue('vrf.rounds', undefined);
+    // }, [setFieldValue, values?.vrf?.campaign]);
+
+    // useSkipEffectOnMount(() => {
+    //     setFieldValue('vrf.rounds', undefined);
+    // }, [setFieldValue, values?.vrf?.vaccine_type]);
```
Comment (beygorghor) [inline plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VaccineRequestForm/VaccineRequestForm.tsx]: 🗑️  ?

File: plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VaccineRequestForm/VaccineRequestForm.tsx [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                                component={DateInput}
+                                disabled={false}
+                            />
+                        </Grid>
+                    </Grid>
+                    <Grid container item xs={12} md={9} lg={6} spacing={2}>
+                        <Grid item xs={12}>
+                            <TextArea
+                                value={values?.vrf?.comment ?? ''}
+                                // errors={errors.comment ? errors.comment : []}
```
Comment (beygorghor) [inline plugins/polio/js/src/domains/VaccineModule/SupplyChain/Details/VaccineRequestForm/VaccineRequestForm.tsx:257]: 🗑️

File: plugins/polio/js/src/domains/VaccineModule/SupplyChain/hooks/validation.ts
Comment (beygorghor) [inline plugins/polio/js/src/domains/VaccineModule/SupplyChain/hooks/validation.ts:1]: 😮

**Review by beygorghor — APPROVED** [review]
Nice job, especially the separate save. 
maybe try to use more keep previous data and avoid screen flickering
Non blocking issues:

Screenshot 2023-12-08 at 17 16 56
- this is not clear those field are required and error is not visible
Screenshot 2023-12-08 at 17 16 40
- i have a grey border
Screenshot 2023-12-08 at 17 19 25
- can we align the icons with the middle of the tab ?
Screenshot 2023-12-08 at 17 20 21
- The spacing is too big on prealert and vars

File: plugins/polio/js/src/components/Inputs/MultiSelect.tsx (hunk 3 above)
Comment (quang-le) [reply plugins/polio/js/src/components/Inputs/MultiSelect.tsx:53]: I'm not sure I should keep this, I was trying something out but didn't use it in the end

**Pull request merged**